File: flutter-local/lib/engine/auth_service.dart
import 'package:flutter/foundation.dart';

import 'data_store.dart';
import 'password_hasher.dart';

/// Manages authentication state and role-based access control for ODS apps.
///
/// ODS Ethos: The framework handles all auth complexity. Builders just add
/// `"roles": ["admin"]` to their spec elements, and AuthService makes it work.
///
/// This service is owned by AppEngine (like DataStore) and provides:
///   - Login/logout session management
///   - Admin setup wizard state
///   - Role-based access checks (the core `hasAccess` method)
///   - User CRUD operations (delegated to DataStore)
class AuthService extends ChangeNotifier {
  final DataStore _dataStore;

  // Session state
  int? _currentUserId;
  String? _currentUsername;
  String? _currentDisplayName;
  List<String> _currentRoles = [];
  bool _isAdminSetUp = false;
  bool _isInitialized = false;

  AuthService(this._dataStore);

  // ---------------------------------------------------------------------------
  // Public getters
  // ---------------------------------------------------------------------------

  bool get isInitialized => _isInitialized;
  bool get isLoggedIn => _currentUserId != null;
  bool get isGuest => _currentUserId == null;
  bool get isAdmin => _currentRoles.contains('admin');
  bool get isAdminSetUp => _isAdminSetUp;

  int? get currentUserId => _currentUserId;
  String get currentUsername => _currentUsername ?? 'guest';
  String get currentDisplayName => _currentDisplayName ?? 'Guest';

  /// Returns the current user's roles. Guests get ['guest'].
  List<String> get currentRoles => isGuest ? const ['guest'] : _currentRoles;

  // ---------------------------------------------------------------------------
  // Core permission check
  // ---------------------------------------------------------------------------

  /// Checks whether the current user has access to an element with the given
  /// role restriction.
  ///
  /// Returns true when:
  ///   - [requiredRoles] is null or empty (no restriction)
  ///   - The current user is an admin (admin bypasses all restrictions)
  ///   - The current user has at least one matching role
  bool hasAccess(List<String>? requiredRoles) {
    if (requiredRoles == null || requiredRoles.isEmpty) return true;
    if (isAdmin) return true;
    return currentRoles.any((r) => requiredRoles.contains(r));
  }

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  /// Initializes the auth service: creates auth tables and checks admin state.
  /// Called by AppEngine.loadSpec() when auth.multiUser is true.
  Future<void> initialize() async {
    await _dataStore.ensureAuthTables();
    _isAdminSetUp = await _dataStore.hasAdminUser();
    _isInitialized = true;
    notifyListeners();
  }

  // ---------------------------------------------------------------------------
  // Authentication operations
  // ---------------------------------------------------------------------------

  /// Attempts to log in with the given credentials.
  /// Returns true on success, false on failure.
  Future<bool> login(String username, String password) async {
    final user = await _dataStore.getUserByUsername(username);
    if (user == null) return false;

    final storedHash = user['password_hash'] as String;
    final salt = user['salt'] as String;

    if (!PasswordHasher.verify(password, salt, storedHash)) return false;

    _currentUserId = user['_id'] as int;
    _currentUsername = user['username'] as String;
    _currentDisplayName = user['display_name'] as String?;
    _currentRoles = await _dataStore.getUserRoles(_currentUserId!);
    notifyListeners();
    return true;
  }

  /// Logs out the current user, reverting to guest state.
  void logout() {
    _currentUserId = null;
    _currentUsername = null;
    _currentDisplayName = null;
    _currentRoles = [];
    notifyListeners();
  }

  /// Creates the initial admin account. Called from the admin setup wizard.
  /// Returns true on success.
  Future<bool> setupAdmin(String username, String password) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(password, salt);

      final userId = await _dataStore.createUser(
        username: username,
        passwordHash: hash,
        salt: salt,
        displayName: username,
      );

      await _dataStore.assignRole(userId, 'admin');
      await _dataStore.assignRole(userId, 'user');

      _isAdminSetUp = true;

      // Auto-login as the new admin.
      _currentUserId = userId;
      _currentUsername = username;
      _currentDisplayName = username;
      _currentRoles = ['admin', 'user'];
      notifyListeners();
      return true;
    } catch (e) {
      debugPrint('ODS AuthService: Admin setup failed: $e');
      return false;
    }
  }

  /// Registers a new user with the given role.
  /// Returns the user ID on success, null on failure.
  Future<int?> registerUser({
    required String username,
    required String password,
    required String role,
    String? displayName,
  }) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(password, salt);

      final userId = await _dataStore.createUser(
        username: username,
        passwordHash: hash,
        salt: salt,
        displayName: displayName ?? username,
      );

      await _dataStore.assignRole(userId, role);
      // All non-guest users also get the 'user' base role.
      if (role != 'user' && role != 'guest') {
        await _dataStore.assignRole(userId, 'user');
      }

      notifyListeners();
      return userId;
    } catch (e) {
      debugPrint('ODS AuthService: Registration failed: $e');
      return null;
    }
  }

  /// Changes the password for a user.
  Future<bool> changePassword(int userId, String newPassword) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(newPassword, salt);
      await _dataStore.updateUserPassword(userId, hash, salt);
      return true;
    } catch (e) {
      debugPrint('ODS AuthService: Password change failed: $e');
      return false;
    }
  }

  /// Returns all users with their roles (admin-only operation).
  Future<List<Map<String, dynamic>>> listUsers() async {
    return _dataStore.listUsers();
  }

  /// Deletes a user by ID.
  Future<void> deleteUser(int userId) async {
    await _dataStore.deleteUser(userId);
    notifyListeners();
  }

  /// Assigns a role to a user.
  Future<void> assignRole(int userId, String role) async {
    await _dataStore.assignRole(userId, role);
    // Refresh current user's roles if they were affected.
    if (userId == _currentUserId) {
      _currentRoles = await _dataStore.getUserRoles(userId);
    }
    notifyListeners();
  }

  /// Removes a role from a user.
  Future<void> removeRole(int userId, String role) async {
    await _dataStore.removeRole(userId, role);
    if (userId == _currentUserId) {
      _currentRoles = await _dataStore.getUserRoles(userId);
    }
    notifyListeners();
  }

  /// Resets the auth service to its initial state. Called on app close.
  void reset() {
    _currentUserId = null;
    _currentUsername = null;
    _currentDisplayName = null;
    _currentRoles = [];
    _isAdminSetUp = false;
    _isInitialized = false;
  }
}

File: react-web/src/engine/auth-service.ts
import type PocketBase from 'pocketbase'
import { warn, error } from './log-service.ts'

/**
 * Authentication and role-based access control using PocketBase auth.
 *
 * ODS Ethos: The framework handles all auth complexity. Builders just add
 * `"roles": ["admin"]` to their spec elements, and AuthService makes it work.
 *
 * PocketBase handles password hashing, tokens, and sessions natively.
 * ODS roles are stored as a JSON array field on the user record.
 *
 * Login uses email as the primary identifier. OAuth2 providers configured
 * in PocketBase are automatically discovered and shown on the login screen.
 */
export class AuthService {
  private pb: PocketBase
  private _isAdminSetUp = false
  private _isInitialized = false
  /** When true, the PocketBase superadmin is running this app — bypass all role checks. */
  private _isSuperAdmin = false
  /** Cached list of OAuth2 providers configured in PocketBase. */
  private _oauthProviders: OAuthProvider[] = []

  constructor(pb: PocketBase) {
    this.pb = pb
  }

  /** Mark that the PocketBase superadmin is operating this app. */
  setSuperAdmin(value: boolean): void {
    this._isSuperAdmin = value
  }

  get isSuperAdmin(): boolean { return this._isSuperAdmin }

  // ---------------------------------------------------------------------------
  // Public getters
  // ---------------------------------------------------------------------------

  get isInitialized(): boolean { return this._isInitialized }
  get isLoggedIn(): boolean { return this._isSuperAdmin || this.pb.authStore.isValid }
  get isGuest(): boolean { return !this.isLoggedIn }

  get currentUserId(): string | undefined {
    return this.pb.authStore.record?.id
  }

  get currentUsername(): string {
    if (this._isSuperAdmin) return 'admin'
    return (this.pb.authStore.record?.['username'] as string) ?? 'guest'
  }

  get currentDisplayName(): string {
    if (this._isSuperAdmin) return 'Admin'
    return (this.pb.authStore.record?.['displayName'] as string)
      ?? (this.pb.authStore.record?.['name'] as string)
      ?? this.currentEmail
      ?? this.currentUsername
  }

  get currentEmail(): string {
    if (this._isSuperAdmin) {
      return (this.pb.authStore.record?.['email'] as string) ?? ''
    }
    return (this.pb.authStore.record?.['email'] as string) ?? ''
  }

  get currentRoles(): string[] {
    if (this._isSuperAdmin) return ['admin', 'user']
    if (this.isGuest) return ['guest']
    const roles = this.pb.authStore.record?.['roles']
    if (Array.isArray(roles)) return roles as string[]
    if (typeof roles === 'string') {
      try { return JSON.parse(roles) } catch { return ['user'] }
    }
    return ['user']
  }

  get isAdmin(): boolean {
    return this.currentRoles.includes('admin')
  }

  get isAdminSetUp(): boolean {
    return this._isAdminSetUp
  }

  /** Returns the list of OAuth2 providers configured in PocketBase. */
  get oauthProviders(): OAuthProvider[] {
    return this._oauthProviders
  }

  // ---------------------------------------------------------------------------
  // Core permission check
  // ---------------------------------------------------------------------------

  hasAccess(requiredRoles: string[] | undefined): boolean {
    if (!requiredRoles || requiredRoles.length === 0) return true
    if (this._isSuperAdmin) return true
    if (this.isAdmin) return true
    return this.currentRoles.some(r => requiredRoles.includes(r))
  }

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  /**
   * Initializes the auth service: discovers OAuth2 providers,
   * checks if an admin user exists.
   */
  async initialize(): Promise<void> {
    // Check if any user with admin role exists.
    // The roles filter may 400 if the field doesn't exist yet — fall back
    // to listing all users and checking manually.
    try {
      let foundAdmin = false
      try {
        const admins = await this.pb.collection('users').getFullList({
          filter: 'roles ~ "admin"',
          requestKey: null,
        })
        foundAdmin = admins.length > 0
      } catch {
        // roles field may not exist — check all users manually
        try {
          const allUsers = await this.pb.collection('users').getFullList({ requestKey: null })
          foundAdmin = allUsers.some((u) => {
            const roles = u['roles']
            if (typeof roles === 'string') return roles.includes('admin')
            if (Array.isArray(roles)) return (roles as string[]).includes('admin')
            return false
          })
        } catch {
          // Users collection may not exist
        }
      }
      this._isAdminSetUp = foundAdmin
    } catch {
      this._isAdminSetUp = false
    }

    // Discover OAuth2 providers configured in PocketBase
    try {
      const methods = await this.pb.collection('users').listAuthMethods({ requestKey: null } as Record<string, unknown>)
      this._oauthProviders = (methods.oauth2?.providers ?? []).map((p) => ({
        name: p.name as string,
        displayName: p.displayName as string ?? p.name as string,
        state: p.state as string,
        authURL: p.authURL as string,
        codeVerifier: p.codeVerifier as string,
        codeChallenge: p.codeChallenge as string,
        codeChallengeMethod: p.codeChallengeMethod as string,
      }))
    } catch {
      this._oauthProviders = []
    }

    this._isInitialized = true
  }

  // ---------------------------------------------------------------------------
  // Authentication operations
  // ---------------------------------------------------------------------------

  /** Attempt to log in with email + password. Returns true on success. */
  async login(email: string, password: string): Promise<boolean> {
    try {
      await this.pb.collection('users').authWithPassword(email, password)
      return true
    } catch {
      return false
    }
  }

  /**
   * Start OAuth2 redirect flow. Saves state to localStorage and redirects
   * the browser to the provider's auth URL. After the user authenticates,
   * PocketBase redirects back to our app where completeOAuth2() finishes
   * the flow.
   */
  async startOAuth2Redirect(providerName: string): Promise<void> {
    try {
      const methods = await this.pb.collection('users').listAuthMethods({ requestKey: null } as Record<string, unknown>)
      const provider = (methods.oauth2?.providers ?? []).find(
        (p) => p.name === providerName,
      )
      if (!provider) {
        warn('AuthService', `OAuth2 provider "${providerName}" not found`)
        return
      }

      // Save provider state for the callback
      localStorage.setItem('ods_oauth2_provider', providerName)
      localStorage.setItem('ods_oauth2_state', provider.state as string)
      localStorage.setItem('ods_oauth2_codeVerifier', provider.codeVerifier as string)
      localStorage.setItem('ods_oauth2_returnUrl', window.location.href)

      // Redirect to provider — use our app's callback URL (not PB's built-in one)
      const redirectUrl = `${window.location.origin}/oauth2-callback`
      const authUrl = (provider.authURL as string) + encodeURIComponent(redirectUrl)

      window.location.href = authUrl
    } catch (e) {
      error('AuthService', 'OAuth2 redirect failed', e)
    }
  }

  /**
   * Complete the OAuth2 flow after the provider redirects back.
   * Extracts the code from URL params and exchanges it for a PB auth token.
   * Returns true on success.
   */
  async completeOAuth2(code: string, state: string): Promise<boolean> {
    try {
      const savedProvider = localStorage.getItem('ods_oauth2_provider') ?? ''
      const savedState = localStorage.getItem('ods_oauth2_state') ?? ''
      const codeVerifier = localStorage.getItem('ods_oauth2_codeVerifier') ?? ''

      if (state !== savedState) {
        warn('AuthService', 'OAuth2 state mismatch')
        return false
      }

      const redirectUrl = `${window.location.origin}/oauth2-callback`

      const result = await this.pb.collection('users').authWithOAuth2Code(
        savedProvider,
        code,
        codeVerifier,
        redirectUrl,
        {},  // createData
        {},  // body
        { requestKey: null },  // prevent auto-cancellation
      )

      // Ensure the OAuth user has roles set (new users won't have them)
      const roles = result.record['roles']
      if (!roles || roles === '[]' || roles === '') {
        await this.pb.collection('users').update(result.record.id, {
          roles: JSON.stringify(['user']),
        })
        await this.pb.collection('users').authRefresh()
      }

      // Clean up localStorage
      localStorage.removeItem('ods_oauth2_provider')
      localStorage.removeItem('ods_oauth2_state')
      localStorage.removeItem('ods_oauth2_codeVerifier')

      return true
    } catch (e) {
      error('AuthService', 'OAuth2 code exchange failed', e)
      return false
    }
  }

  /** Get the saved return URL after OAuth2 callback. */
  static getOAuth2ReturnUrl(): string | null {
    const url = localStorage.getItem('ods_oauth2_returnUrl')
    localStorage.removeItem('ods_oauth2_returnUrl')
    return url
  }

  /** Log out the current user. */
  logout(): void {
    this.pb.authStore.clear()
  }

  /**
   * Create the initial admin account. Called from the admin setup wizard.
   * Creates a PocketBase user with admin + user roles.
   */
  async setupAdmin(email: string, password: string, displayName?: string): Promise<boolean> {
    try {
      // Reject if email matches PocketBase superadmin — they are separate identity stores
      const pbAdminEmail = (this.pb.authStore.record?.['email'] as string) ?? ''
      if (pbAdminEmail && email.toLowerCase() === pbAdminEmail.toLowerCase()) {
        warn('AuthService', 'Cannot create app user with same email as PocketBase superadmin')
        return false
      }

      // Generate a username from the email (before @)
      const username = email.split('@')[0].replace(/[^\w]/g, '_').toLowerCase()

      await this.pb.collection('users').create({
        username,
        password,
        passwordConfirm: password,
        email,
        displayName: displayName ?? username,
        roles: JSON.stringify(['admin', 'user']),
      })

      // Auto-login as the new admin
      await this.pb.collection('users').authWithPassword(email, password)
      this._isAdminSetUp = true
      return true
    } catch (e) {
      error('AuthService', 'Admin setup failed', e)
      return false
    }
  }

  /** Register a new user with the given role. Returns user ID on success. */
  async registerUser(params: {
    email: string
    password: string
    role: string
    displayName?: string
  }): Promise<string | null> {
    try {
      // Reject if email matches PocketBase superadmin
      const pbAdminEmail = (this.pb.authStore.record?.['email'] as string) ?? ''
      if (pbAdminEmail && params.email.toLowerCase() === pbAdminEmail.toLowerCase()) {
        warn('AuthService', 'Cannot register with PocketBase superadmin email')
        return null
      }

      const roles = [params.role]
      if (params.role !== 'user' && params.role !== 'guest') {
        roles.push('user')
      }

      // Generate username from email
      const username = params.email.split('@')[0].replace(/[^\w]/g, '_').toLowerCase()
        + '_' + Math.random().toString(36).slice(2, 6)

      const record = await this.pb.collection('users').create({
        username,
        password: params.password,
        passwordConfirm: params.password,
        email: params.email,
        displayName: params.displayName ?? params.email.split('@')[0],
        roles: JSON.stringify(roles),
      })

      return record.id
    } catch (e) {
      error('AuthService', 'Registration failed', e)
      return null
    }
  }

  /**
   * Validate a password against PocketBase requirements.
   * Returns null if valid, or an error message string.
   */
  static validatePassword(password: string): string | null {
    if (password.length < 8) return 'Password must be at least 8 characters.'
    if (password.length > 72) return 'Password must be 72 characters or fewer.'
    return null
  }

  /** Change password for a user. */
  async changePassword(userId: string, newPassword: string): Promise<boolean> {
    try {
      await this.pb.collection('users').update(userId, {
        password: newPassword,
        passwordConfirm: newPassword,
      })
      return true
    } catch {
      return false
    }
  }

  /** List all users (admin operation). */
  async listUsers(): Promise<Record<string, unknown>[]> {
    try {
      const records = await this.pb.collection('users').getFullList({ sort: 'created', requestKey: null })
      return records.map(r => ({
        _id: r.id,
        username: r['username'],
        email: r['email'] ?? '',
        displayName: r['displayName'] ?? r['name'] ?? r['email'] ?? r['username'],
        roles: (() => {
          const roles = r['roles']
          if (Array.isArray(roles)) return roles
          if (typeof roles === 'string') {
            try { return JSON.parse(roles) } catch { return [] }
          }
          return []
        })(),
        _createdAt: r.created,
      }))
    } catch {
      return []
    }
  }

  /** Delete a user by ID. */
  async deleteUser(userId: string): Promise<void> {
    await this.pb.collection('users').delete(userId)
  }

  /** Assign a role to a user. */
  async assignRole(userId: string, role: string): Promise<void> {
    const user = await this.pb.collection('users').getOne(userId)
    let roles: string[] = []
    try { roles = JSON.parse(user['roles'] as string) } catch { /* empty */ }
    if (!roles.includes(role)) {
      roles.push(role)
      await this.pb.collection('users').update(userId, {
        roles: JSON.stringify(roles),
      })
    }
  }

  /** Remove a role from a user. */
  async removeRole(userId: string, role: string): Promise<void> {
    const user = await this.pb.collection('users').getOne(userId)
    let roles: string[] = []
    try { roles = JSON.parse(user['roles'] as string) } catch { /* empty */ }
    roles = roles.filter(r => r !== role)
    await this.pb.collection('users').update(userId, {
      roles: JSON.stringify(roles),
    })
  }

  /** Reset to initial state. */
  reset(): void {
    this.pb.authStore.clear()
    this._isAdminSetUp = false
    this._isInitialized = false
    this._isSuperAdmin = false
    this._oauthProviders = []
  }
}

/** Describes an OAuth2 provider discovered from PocketBase. */
export interface OAuthProvider {
  name: string
  displayName: string
  state: string
  authURL: string
  codeVerifier: string
  codeChallenge: string
  codeChallengeMethod: string
}
